### PR TITLE
add 'non-ci-build' feature

### LINKED ISSUE #226: Use Universum for non-CI builds
remove unnecessary steps:
* 1. Preparing repository
* 4. Reporting build start
* 8. Finalizing

add flag to make collecting artifacts optional
summarised build results are still required

This mode should be enabled by one flag.
In addition if this mode is enabled next configuration should be used by default:
`--vcs-type=none --server-type=local --out-type=term --launcher-output=console --clean-build --file-source-dir=.`

Originally created on Wed, 12 Dec 2018 14:35:23 +0200

### PR DESCRIPTION
**Pull request by i-savynikh**
# Description
Non CI build feature implementation

Resolves #226

# How Has This Been Tested?
run tests
run universum bin with 'nonci' command line argument

### CODE REVIEW

File: tests/deployment_utils.py [hunk 1]
```diff
@@ -231,10 +232,16 @@ def run(self, config, force_installed=False, vcs_type="none",
     def clean_artifacts(self):
         self.environment.assert_successful_execution("rm -rf '{}'".format(self.artifact_dir))
 
+    def _create_temp_config(self, config):
+        file_path = os.path.join(self.working_dir, "temp_config.py")
+        with open(file_path, 'wb+') as f:
+            f.write(config)
+            f.close()
+        return file_path
```
Comment (k-dovgan) [inline tests/deployment_utils.py]: We tend to declare functions before we use them, for better readability

File: tests/deployment_utils.py [hunk 2]
```diff
@@ -205,20 +206,20 @@ def _vcs_args(self, vcs_type):
...
         # We cannot collect coverage from installed module, so we run it only if specifically told so
         if force_installed:
             cmd = "universum"
 
-        cmd += " -lcp '{}'".format(config_file) \
-               + self._basic_args() + self._vcs_args(vcs_type) + additional_parameters
+        if self.command == 'nonci':
+            cmd += ' nonci'
+        else:
+            cmd += self._vcs_args(vcs_type)
```
Comment (k-dovgan) [inline tests/deployment_utils.py]: This code seems to allow using subcommands other then 'nonci'; but passing any other command will lead to wrong argument set and failure.
I'd rather implement full support of subcommands or pass 'nonci' as flag instead of multi-value string.

File: tests/deployment_utils.py [hunk 3]
```diff
@@ -159,10 +159,11 @@ def local_sources(tmpdir):
 
 
 class UniversumRunner(object):
-    def __init__(self, perforce_workspace, git_client, local_sources):
+    def __init__(self, perforce_workspace, git_client, local_sources, command=None):
```
Comment (k-dovgan) [inline tests/deployment_utils.py]: Isn't this 'subcommand'?

File: doc/args_nonci.rst [hunk 4]
```diff
@@ -0,0 +1,21 @@
+:orphan:
+
+Non-CI command line
+-------------------
+
+An aim of this command is using 'Universum' as add-on for build system.
```
[the remark this answers is not in the document]
Comment (i-savynikh) [reply doc/args_nonci.rst]: However poll and submit are described in exactly this manner

File: doc/args_nonci.rst [hunk 5]
```diff
@@ -0,0 +1,21 @@
+:orphan:
+
+Non-CI subcommand
+-------------------
+
+Purpose of this mode is running 'Universum' configurations locally.
+An aim of this mode is using Universum as part of build system.
+When 'Universum' running in this mode next CI-related features disabled:
```
Comment (i-keliukh) [inline doc/args_nonci.rst]: "next" must be changed to "the following"

**Comment by i-keliukh** [comment]
Also, do you plan to add tests to check nonci functionality specifically?

File: doc/args_nonci.rst (hunk 5 above)
Comment (i-keliukh) [reply doc/args_nonci.rst]: And the comma is missing after the wors "mode".

**Comment by i-savynikh** [comment]
> 
> 
> Also, do you plan to add tests to check nonci functionality specifically?

yes

File: tests/test_code_report.py [hunk 6]
```diff
@@ -2,6 +2,7 @@
 # -*- coding: UTF-8 -*-
 
 import re
+import pytest
```
Comment (k-dovgan) [inline tests/test_code_report.py]: Why is this needed?

File: tests/deployment_utils.py [hunk 7]
```diff
@@ -246,6 +253,14 @@ def universum_runner(execution_environment, runner_without_environment):
     yield runner_without_environment
 
 
+@pytest.fixture(params=(False, True))
+def universum_runner_nonci(execution_environment, runner_without_environment, request):
```
Comment (k-dovgan) [inline tests/deployment_utils.py]: The name of this fixture implies it is used only for 'nonci' mode, while, as I see in code further, it actually is used both in basic and nonci modes.

File: tests/deployment_utils.py [hunk 8]
```diff
@@ -173,7 +174,7 @@ def __init__(self, perforce_workspace, git_client, local_sources):
     def set_environment(self, execution_environment):
         self.environment = execution_environment
         self.working_dir = self.environment.get_working_directory()
-        self.project_root = os.path.join(self.working_dir, "temp")
+        self.project_root = self.working_dir if self.nonci else os.path.join(self.working_dir, "temp")
```
Comment (k-dovgan) [inline tests/deployment_utils.py]: Do we need a one-line if-statement here?

File: doc/args_nonci.rst [hunk 9]
```diff
@@ -0,0 +1,21 @@
+:orphan:
+
+Non-CI subcommand
+-------------------
+
```
Comment (k-dovgan) [inline doc/args_nonci.rst]: I still don't see a direct answer to the question 'What is this?'

File: doc/args_nonci.rst [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+Purpose of this mode is running 'Universum' configurations locally.
+An aim of this mode is using Universum as part of build system.
```
Comment (k-dovgan) [inline doc/args_nonci.rst]: Word 'aim' cannot be used in this context. Please rephrase, e.g. "This mode was designed to use Universum as a part..."

File: doc/args_nonci.rst (hunk 9 above)
Comment (i-savynikh) [reply doc/args_nonci.rst]: Sorry, I don't know what I can add here to fulfil all requirements

File: tests/deployment_utils.py (hunk 8 above)
Comment (i-savynikh) [reply tests/deployment_utils.py]: yes

File: doc/args_nonci.rst [hunk 11]
```diff
@@ -0,0 +1,21 @@
+:orphan:
+
+Non-CI subcommand
+-------------------
+
+The purpose of this mode is running 'Universum' locally.
+This mode is designed to use Universum as part of build system.
+When 'Universum' running in this mode, the following CI-related features disabled:
```
Comment (i-keliukh) [inline doc/args_nonci.rst]: features *are* disabled

File: doc/args_nonci.rst [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+- reporting (to review board for example)
```
Comment (i-keliukh) [inline doc/args_nonci.rst]: - remove brackets
- we do not support the Review Board

File: doc/args_nonci.rst [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+- strict verifying of artifacts before build
```
Comment (i-keliukh) [inline doc/args_nonci.rst]: It is not clear from this line, what is this verification

File: doc/args_nonci.rst [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+- running build scripts within a temporary project directory when VCS is none
```
Comment (i-keliukh) [inline doc/args_nonci.rst]: Totally unclear item. Even I don't understand it.

File: doc/args_nonci.rst [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+
+However you still have pretty build result summary.
```
Comment (i-keliukh) [inline doc/args_nonci.rst]: We need different name for result summary. Is it mentioned elsewhere in docs?
